controllers/volumereplicationgroup_controller.go

```diff
@@ -364,3 +360,126 @@ func (v *VolumeReplicationGroupReconciler) SetupWithManager(mgr ctrl.Manager) er
+//   VRG to backup/replicate the PV to an object store and has
+// 	 other means to replicate PV metadata and is likely using
+//   the VRG as a single control point to manipulate replication
+// 	 of a group of VRs and thus a group of PVs.  However, log
+//   a warning once per VRG
+func (v *VolumeReplicationGroupReconciler) uploadPVofPVC(
+	ctx context.Context,
+	volRepGroup *ramendrv1alpha1.VolumeReplicationGroup,
+	pvc *corev1.PersistentVolumeClaim) error {
+	vrgName := volRepGroup.Name
+
+	s3Endpoint := volRepGroup.Spec.S3Endpoint
+	if s3EndpointIsUnconfiguredOrInvalid(s3Endpoint, vrgName) {
+		return nil
+	}
+
+	s3Conn, err := connectToS3Endpoint(ctx, v,
+		s3Endpoint,
+		types.NamespacedName{ /* secretName */
+			Name:      volRepGroup.Spec.S3SecretName,
+			Namespace: volRepGroup.Namespace,
+		},
+		vrgName, /* debugTag */
+	)
+	if err != nil {
+		return err
+	}
+
+	pv := corev1.PersistentVolume{}
+	volumeName := pvc.Spec.VolumeName
+	pvObjectKey := client.ObjectKey{
+		Name: volumeName,
+	}
+
+	if err := v.Get(ctx, pvObjectKey, &pv); err != nil {
+		return errorswrapper.Wrapf(err, "failed to get upload PV %v of VRG %v", volumeName, vrgName)
+	}
+
+	// Create the bucket, without caching or assuming its existence; use the VRG name
+	// as the bucket name
+	bucket := vrgName
+	if err := s3Conn.createBucket(bucket); err != nil {
+		log.Errorf("unable to create bucket %v for VRG %v", bucket, vrgName)
+
+		return err
+	}
+
+	if err := s3Conn.uploadPV(bucket, pv.Name, pv); err != nil {
+		return err
+	}
+
+	if err := s3Conn.verifyPVUpload(bucket, pv.Name, pv); err != nil {
+		return err
+	}
+
+	// Remove after Benamar uses this method in AVR
+	if _, err := s3Conn.downloadPVs(bucket); err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (v *VolumeReplicationGroupReconciler) fetchVolRepGroup(
+	ctx context.Context, req ctrl.Request) (
+	*ramendrv1alpha1.VolumeReplicationGroup, error) {
+	volRepGroup := &ramendrv1alpha1.VolumeReplicationGroup{}
+	if err := v.Get(ctx, req.NamespacedName, volRepGroup); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			log.Info("VolumeReplicationGroup resource not found. Ignoring since object must have been deleted: ",
+				req.NamespacedName)
+
+			return nil, nil
+		}
+
+		// Error reading the object
+		log.Error(err, "Failed to get VolumeReplicationGroup")
+
+		return nil, errorswrapper.Wrapf(err, "Failed to get VolumeReplicationGroup %s", req.NamespacedName)
+	}
+
+	return volRepGroup, nil
+}
+
+// If the given s3 endpoint for the given VRG is unconfigured or
+// invalid, log a warning message, but only once
+func s3EndpointIsUnconfiguredOrInvalid(s3Endpoint, vrgName string) (
+	invalid bool) {
+	invalid = true
+
+	if prevEndpoint, prevWarned := s3Warning[vrgName]; prevWarned &&
```

Comment:
Why do you need to cache in the s3Warning?  Is it because you only want to see the warning message once in the logs? I think this block is not needed.

Reply:
Yes, that was reason, especially with a view of supporting an unconfigured S3Endpoint.  We should discuss whether making S3Endpoint optional is of value for the case I cited earlier.